accepted/00000-ansible-integration.md

```diff
@@ -0,0 +1,320 @@
+
+Here we're **not** primarely focused on building a new Ansible managed infrastructure from scratch using Uyuni.
+
+NOTE: Currently Ansible is not shipped in SLE15 any SP. In order to allow an SLE Ansible control node, we would need to ship Ansible in probably in SLE15 client tools.
+
+2. Using the Uyuni server as the Ansible control node:
+
+On this approach, the Uyuni server is our Ansible control node. We would need to provide the Ansible package to the SUSE Manager server channel.
+
+Since the Ansible inventory in the Uyuni server would be initially empty, default at `/etc/ansible/hosts`, the user would add their Ansible managed hosts to the inventory and will take care of deploying the Ansible SSH keys to those systems, in order for Ansible to work. (For already registered systems, deploying the SSH key could be automated somehow)
+
+Assuming the Ansible inventory is properly created and SSH keys are deployed, the user can operate it from the command line, but Uyuni can also operate it via Salt runner call (since we're on the Uyuni server and not in a minion). Example:
+
+```console
+# salt-run salt.cmd ansible.playbooks name=example_playbook.yml
+```
+
+This means, playbooks can be triggered via CLI by the user, but also via "salt-api" so Uyuni is also easily able to trigger playbooks executions.
+
+Once we have a "Playbook catalog" in the UI, and Uyuni would trigger playbook executions we would need to implement of course a new type of action: `ApplyPlaybook`, which should expose the playbook to the Salt "file_roots" so it's available for Ansible when running.
+
+3. Adding an Ansible control node from a registered minion:
+
+This scenario refers to the posibility of adding an Ansible control node from a registered system. Similarly to the previous scenarios but in this case, the Ansible control node is a registered system in Uyuni.
+
+### ansible-gatherer
+
+This is the component that takes care of collecting the inventory, host and playbooks from a given Ansible control node.
+
+Therefore, there might be more than one Ansible control node host, even the Uyuni server, proxy or a registered system, could act at some point as an Ansible control node. Since data sources, and type of sources (e.g. single host, AWX API) might be multiple, this RFC proposes an approach similarly to what Uyuni does for handling "Virtual Host Managers (VHM)". This means, using a Python tool, in this case called something like "ansible-gatherer", which is plugin-based (so easily allows implementing different sources of Ansible inventories). This tool would be called via an Uyuni Java schedule, like the "virtual-host-gatherer", passing the necessary information (parameters, type of host, etc) to reach the Ansible control node and collect the necessary information.
```

Comment:
I see the idea of rebranding the VHM to something more generic as interesting and worth to consider.

Pros:
- Less components in Uyuni. Probably less code lines in general.
- Reduce a bit the Python implementation (just adding new plugins to "virtual-host-gatherer"). Most of the implementation effors goes into Java / UI.
- Reuse of Taskomatic tasks, some parts of the UI and XMLRPC API.

Cons:
- Adds more complexity on the Java backend that receives the JSON output from "virtual-host-gatherer":  The current Java backend for VHM is basically meant for gathering Virtual Hosts instances. The JSON output produced by "virtual-host-gatherer" is completely different, and with a different purpose, depending if you're, for example gathering virtual instances (where you get names + uuids) or gathering Ansible nodes from an Ansible inventory (not uuid, playbooks, etc). The Java backend for VHM would need to be adapted to check which "plugin" has been triggered and then parse the output accordingly.
- Having the same schedule for gathering from Ansible that for any other VHM might not be necessarely what a user wants.
- While some of the UI can be reused, also new UI would be anywa y required to visualize an Ansible control node (which is not the same UI that we expect when visualizing a VHM node).
- Completely isolate Ansible gathering from VHM sounds simpler.

Reply:
Pros: I agree.

Cons: 
 - I agree about the same schedule argument
 - I do not see the Java JSON parsing part to be significantly different, in terms of complexity, in the two cases. No matter what we do we will always need a new parser of a completely different JSON structure
 - Similarly, Ansible-specific UI parts will have to be new in any case
 - I agree this adds a bit of complexity, but...

Additonal elements I see are:
 - we will need to rebrand the VHM in any case, as it's long past the point of being an accurate name. Doing so in this occasion is "taking two birds with one shot". We will later save that effort
 - we would establish a standard way to add systems to SUSE Manager in addition to onboarding. Anything else we would add in future would have a clear place to fit in, and clear example to copypaste from. It will be closer to a Salt-style plugin infrastructure (for this particular aspect) - a more modular design
 - I also see it could it be easier on user experience in general to have a unifying concept for any "external things". Whatever you want to add to SUMa, you go to the one page to configure details on how to access such an "external thing", and know there is one task to keep it current

How does that sound?

Reply:
I see your point and I like the idea of rebrading the VHM to something more generic, but honestly I think that would require more time and will deserve its own RFC.

Rebrading VHM means, not only code/UI changes (probably also package name changes) but also lot of changes in the documentation.

At this point in time, I would prefer to start the Ansible integration without unifying it with the VHM. Of course, we can later work on a "Rebrading VHM RFC", where I think most of the effors invested in the backend of Java for parsing the output from `ansible-gatherer` can be also reused in a future generic importer. Same for the UI.

What do you think?

Reply:
@moio we already discussed the idea of making VHM more powerful and integrated and I am all for it but that's a separate RFC, one for the future. Let's focus on the problem at hand: Ansible.

Reply:
I am not suggesting to make the VHM any more powerful than it is now, I am just suggesting to take this occasion to rename it, and put the Ansible integration in that place (UI-wise and code-wise) instead of someplace new.

So far I did not yet see a convincing argument why this would turn out to be substantially more complex. Even docs-wise we are really talking about ~10 files and a few `sed` commands.

I could even agree Uyuni-wise and RFC-wise it would make sense to postpone this, but downstream-wise we do not want such a UI change (even if it's just mere renaming) before a major version, so now it is a good time to do it.

That would make the overall architecture easier to understand and maintain, instead of adding yet another ad-hoc mechanism.

Reply:
While I can understand all the point of view explained so far, I'd like to put my thought on this, and I will do it by "cleaning up the table" and start thinking from scratch.

From a general perspective Uyuni is capable to add systems and keep track of them: the main purpose of doing this is to manage them on the configuration and many other parts, and these systems go through the onboarding process. Let's call these systems the **managed systems**.
That said, Uyuni becomes a core application to manage all the systems in a data center, so "core" that it comes by itself that Uyuni is also an _inventory_ tool of everything available in the data center. Now, Uyuni has also a nice feature to **just keep tracking** of other systems in the data center, the ones Uyuni is not there to **manage** them. Let's call these systems the **self systems**. All the latter are the ones gathered by the VHM.

Here it chimes in our Ansible integration: what are we going to implement? IIUC we will let Uyuni be aware of the existing Ansible systems, but we are definitely not going to manage them their configuration through the onboarding process etc. Therefore they will just take part of the **self systems** bunch mentioned before.

Where all of this is going?
I am trying to abstract the thinking and reaching the user POV, and I want to make sure it is clear where things are and what they are meant for. I (as a user) will fork my searching biased on either **registered systems** because I want to manage their configuration **or** I will look for **self systems** because I need to do whatever I can do with them, but not configuration for sure. The clearest way of representing this is to append the Ansible bunch of systems under the VHM umbrella, but definitely renaming the umbrella to a generic one. Unless I am missing something huge, an Ansible system will just be a **type** of the self systems, like Virtual Host Manager are, and the approach to them is the same.

Why should we not distinguish Ansible as a top level menu entry?
Because we are not trying to distinguish a system by its nature, we want to keep the Uyuni perspective, we are here "to manage or not to manage" first. Once we fork the initial step, we can move the next one in the workflow. Otherwise spaghetti connections are to come inside the user mind, therefore in the code logic.

Looking for a first small initial step to add the integration but not refactoring the whole VHM?
I believe an Ansible system is just a sub-type of the gathered-not-managed systems, but I don't really know the difference from refactoring and including the Ansible gathering withing the current VHM code, or just implementing a separate code path and append it, so I will let this up to who knows this part better. If the assumption of adding a separate code path to gather Ansible systems is fine, I can propose a short-term solution (which is meant to be temporary, the final one will be the whole refactoring): we could wrap the two types under an abstract concept level but keeping them distinguished, in order to let untouched the VHM part. But again, the assumption must be right in the light of the final scope of the whole refactoring.

Before:
`Systems > Virtual Host Managers`

After:
`Systems > Self Systems > Virtual Host Managers`
`Systems > Self Systems > Ansible`

PS: "Self Systems" is just for consistency of my comment, it can be "Autonomous Systems" or "Not Manage Systems" or something else.

Reply:
@ncounter we are going to manage those systems. They are not foreign systems.

Reply:
> @ncounter we are going to manage those systems. They are not foreign systems.

This is not what the initial motivation describes:

>This RFC is more focused on allowing users to import their existing Ansible environments into Uyuni, allow some basic operations, coexistence of Salt and Ansible managed infrastructure in Uyuni, eventually apply playbooks and also to provide an easy way to transition from Ansible to a fully-featured Salt minion managed system in Uyuni, rather than making Uyuni a top-featured UI to build your Ansible infrastructure from scratch.

We will not manage them "by Ansible". We will import them as they are (`allowing users to import their existing Ansible environments into Uyuni`). If the user what to manage them instead, the Salt path will occur (`to transition from Ansible to a fully-featured Salt minion managed system in Uyuni`). Until it is just an Ansible system with no Salt configuration, it should be treated as a foreign/not-managed system. Once the transition is completed, it should end up in the System List as other clients, as a Salt client. Of course the Ansible property should be visible, but it's just a property of the system, we are not handling it by Ansible.

And as mentioned in this section as well `this RFC proposes an approach similarly to what Uyuni does for handling "Virtual Host Managers (VHM)"` because they are "gathered" systems. The question is whether we want to add yet-another-gatherer for a different type, or refactor the VHM and gather multiple types of systems with the same logic, which sounds much more inline with the current picture.

Reply:
@ncounter Step by step, we are going to:

1. Onboard Ansible controllers and gather inventory (Ansible client machines, final leaf in the hierarchy)
2. Gather what playbooks the Ansible controllers can apply on those Ansible client machines
3. Allow running those playbooks from the Uyuni Server instead of the Ansible controller (the Uyuni Server will drive the Ansible controller, not execute directly)
4. Allow adding Ansible playbooks to configuration channels. Mayb we'll call that "Ansible configuration channels", maybe we'll just add them to the current configuration channels. To be determined.
5. Allow adding playbooks directly from the Uyuni Server WebUI so that users do not need to go to the Ansible controller at all
6. Allow editing playbooks directly in the Uyuni Server WebUI, the same way we can edit Salt states

If that is not management, then I don't what it is.

The fact we will use Salt+AnsibleGate to drive Ansible is just a technical detail.

Reply:
@ncounter let me offer a possible clarification/way to look at this - as of today, there's two ways to have a system show up in the Systems list: either via registration, or via an "import" from the VHM. Those are just the two ways users have at their disposal to add a system to Uyuni.

The fact that the _current implementation_ of VHM exclusively imports systems which have very limited functionality compared to the registered ones I'd consider more about a historical consideration than a design one. Abstracting from the Ansible discussion here, it's not set in stone that systems coming somewhere other than a registration have to be "underpowered" in any way. In principle, we could very well add new imported system types which end up being well featured, or fully featured, or (stretching the concept) even have more features than anything we have been registering to this day. It's just a matter of having some API endpoint to discover them instead of a bootstrap script.

What I am trying to say is that the crucial difference from a UX perspective could be how the system "is originated in Uyuni" - to offer the two entrance gates in a consistent way. Once a system is born, it's a system like all others, with a set of available features determined by its System Type (formerly "entitlement"), which we can use as filtering conditions just like all others.

In that light, the level of feature-completeness for Ansible clients becomes less relevant. Today it can be basic, tomorrow it might become super-advanced, but this is pretty much transparent to the onboarding procedure.

Does that sound any clearer?